shachain/producer.go

```diff
@@ -0,0 +1,72 @@
+package shachain
+
+import (
+	"github.com/roasbeef/btcd/wire"
+)
+
+// SecretProducer is an interface which serves as an abstraction over data
+// structure responsible for effecient generating the secrets by given index.
+// The generation of secrets should be made in such way that secret store
+// might efficiently store and retrieve the secrets.
+type SecretProducer interface {
+	// Produce secret by given index.
+	Produce(uint64) (*wire.ShaHash, error)
+
+	// ToBytes convert producer to the binary representation.
+	ToBytes() ([]byte, error)
+}
+
+// ShaChainProducer implementation of SecretProducer. This version of shachain
+// sligtly changed in terms of method naming. Initial concept might be found
+// here:
+// https://github.com/rustyrussell/ccan/blob/master/ccan/crypto/shachain/design.txt
+type ShaChainProducer struct {
+	// root is the element from which we may generate all hashes which
+	// corresponds to the index domain [281474976710655,0].
+	root *Element
+}
+
+// A compile time check to ensure ShaChainProducer implements the SecretProducer
+// interface.
+var _ SecretProducer = (*ShaChainProducer)(nil)
+
+// NewShaChainProducer create new instance of shachain producer.
+func NewShaChainProducer(root *wire.ShaHash) *ShaChainProducer {
+	return &ShaChainProducer{&Element{
+		index: rootIndex,
+		hash:  root,
+	}}
+}
+
+// NewShaChainSender recretes instance of shachain producer by the given
```

Comment:
The leading type name in this comment is incorrect.

Additionally, I think this comment should be reworded to: 
> NewShaChainProducerFromBytes deserlzalies an instance of a ShaChainProducer encoded in the passed byte slice, returning a fully initialize instance of a ShaChainProducer